src/Uri.php

```diff
@@ -594,6 +598,25 @@ private function filterPath(string $path) : string
         return '/' . ltrim($path, '/');
     }
 
+    /**
+     * Encode invalid UTF-8 characters in given string. All other characters are unchanged.
+     */
+    private function filterInvalidUtf8(string $string) : string
+    {
+        if (preg_match('//u', $string)) {
```

Comment:
I don't understand what this is doing, exactly.

I've tried it locally with a variety of strings from your query provider below, and every single one of them results in a positive match, which means the logic below never gets hit. Maybe you're not testing _invalid_ UTF-8 characters anywhere?

Reply:
How come?

See: https://3v4l.org/MEYgF

```php
var_dump(preg_match('//u', "\x21\x92"));
```

result:
```
bool(false)
```

Somehow test I've added works.

Reply:
Turned out I was using single quotes instead of double quotes when I was doing my tests here. Once I changed the quoting style, I was able to observe the same behaviour finally.

Might be good to drop a note in indicating what this is doing, though, so future contributors/maintainers know the purpose.